src/compas/datastructures/mesh/duality.py

```diff
@@ -54,28 +70,53 @@ def mesh_dual(mesh, cls=None, boundary=0):
 
         if not boundary:
             continue
-        if vertex not in outer or len(faces) <= 1:
+
+        if boundary > 3:
+            raise ValueError("edge mode from 0 to 3!")
+
+        if vertex not in outer or len(faces) < 1:
             continue
 
-        nbr_vertices = reversed(mesh.vertex_neighbors(vertex, ordered=True))
-        boundary_vertices = faces
+        boundary_fids = faces[:]
+        current_face = vertex
+        corner_count = 0
+        edge_count = 0
+
+        for nbr_vertex in reversed(mesh.vertex_neighbors(vertex, ordered=True)):
+
+            if not mesh.is_edge_on_boundary(vertex, nbr_vertex):
+                continue
+            pt = mesh.edge_midpoint(vertex, nbr_vertex)
+
+            if num_faces not in vertex_xyz and len(faces) == 1 and corner_count == 0 and (boundary == 2 or boundary == 3):
+                vertex_xyz[num_faces] = mesh.vertex_coordinates(vertex)
+                current_face = num_faces
+                num_faces += 1
+                corner_count += 1
+
+            if num_faces not in vertex_xyz and len(faces) != 1 and edge_count == 0 and boundary == 3:
+                vertex_xyz[num_faces] = mesh.vertex_coordinates(vertex)
+                current_face = num_faces
+                num_faces += 1
+                edge_count += 1
 
-        for nbr_vertex in nbr_vertices:
+            if num_faces not in vertex_xyz and ((vertex, nbr_vertex) not in edge_vertex and
+                                                (nbr_vertex, vertex) not in edge_vertex):
 
-            if mesh.is_edge_on_boundary(vertex, nbr_vertex):
-                pt = mesh.edge_midpoint(vertex, nbr_vertex)
+                vertex_xyz[num_faces] = pt
+                edge_vertex[vertex, nbr_vertex] = edge_vertex[nbr_vertex, vertex] = num_faces
+                boundary_fids.append(num_faces)
+                num_faces += 1
+            else:
+                boundary_fids.append(edge_vertex[vertex, nbr_vertex])
 
-                if geometric_key(pt) not in boundary_xyz and num_faces not in vertex_xyz:
-                    vertex_xyz[num_faces] = pt
-                    boundary_xyz[geometric_key(pt)] = num_faces
-                    num_faces += 1
+        if vertex in outer and len(faces) == 1 and (boundary == 2 or boundary == 3):
+            boundary_fids.insert(len(faces) + 1, current_face)
```

Comment:
why are we making this so complicated?

Reply:
I tried to make this dual mesh as general and simple as possible. For example, we have to consider some special cases when mesh face id is not starting from 0 and is a bit random.